corva/models/stream.py

```diff
@@ -1,27 +1,55 @@
-from typing import Any, Dict, List, Optional
+from typing import Dict, List, Optional
 
-from corva.models.base import BaseContext, BaseEventData
+from corva.models.base import BaseContext, BaseEventData, ListEvent
 from corva.state.redis_state import RedisState
 
 
 class StreamContext(BaseContext):
     state: RedisState
 
 
-class Record(BaseEventData):
-    timestamp: Optional[int] = None
-    asset_id: int
-    company_id: int
-    version: int
-    data: Dict[str, Any]
-    measured_depth: Optional[float] = None
-    collection: str
+class StreamEventData(BaseEventData):
+    class Record(BaseEventData):
+        class Data(BaseEventData):
+            hole_depth: Optional[float] = None
+            weight_on_bit: Optional[int] = None
+            state: Optional[str] = None
 
+        timestamp: Optional[int] = None
+        asset_id: int
+        company_id: int
+        version: int
+        measured_depth: Optional[float] = None
+        collection: str
+        data: Data
 
-class StreamEventData(BaseEventData):
+    class Metadata(BaseEventData):
+        class AppData(BaseEventData):
+            app_connection_id: int
+
+        app_stream_id: int
+        apps: Dict[str, AppData]
+
+    app_key: Optional[str] = None
     records: List[Record]
-    metadata: Dict[str, Any]
-    asset_id: int
-    app_connection_id: int
-    app_stream_id: int
-    is_completed: bool
+    metadata: Metadata
+
+    @property
+    def asset_id(self) -> int:
+        return self.records[0].asset_id
+
+    @property
+    def app_connection_id(self) -> int:
+        return self.metadata.apps[self.app_key].app_connection_id
+
+    @property
+    def app_stream_id(self) -> int:
+        return self.metadata.app_stream_id
+
+    @property
+    def is_completed(self) -> bool:
+        return self.records[-1].collection == 'wits.completed'
```

Comment:
Wrap this with a try so we don't have an IndexError

Reply:
reworked the logic